docs/design/yarn-v1.md

```diff
@@ -0,0 +1,169 @@
+```
+
+##### Git Package
+
+```txt
+pkg:npm/package-name@1.0.0?vcs_url=git%2Bhttps://example.com/namespace/package-name.git%0101010101010101010101010101010101010101
+```
+
+##### Workspace/File Package
+
+```txt
+pkg:npm/package-name@1.0.0?vcs_url=git%2Bhttps://example.com/namespace/package-name.git%0101010101010101010101010101010101010101#subpath
+```
+
+### Avoiding Arbitrary Code Execution
+Yarn 1.x appears to honor the ```--ignore-scripts``` option for the install command. It can also be configured via environment variable.
+Used https://github.com/chmeliik/js-lifecycle-scripts/tree/yarn to verify.
+
+### Prefetch
+The offline mirror in the request output directory will be populated by running the following Yarn command in the project root:
+
+    yarn install --non-interactive --frozen-lockfile --disable-pnp --no-default-rc --ignore-engines
+
+Command: [yarn install](https://classic.yarnpkg.com/en/docs/cli/install)
+
+Options:
+ - non-interactive: Disable interactive prompts
+ - frozen-lockfile: Fail if yarn.lock updates are needed
+ - disable-pnp: Disable Plug'n'Play installation
+ - no-default-rc: Prevent Yarn from automatically detecting .yarnrc and .npmrc files
+ - ignore-engines: Ignore the engine node version during the prefetch
+
+ The behavior of Yarn can also be configured using select environment variables. The following ones should be set for the prefetch:
+
+    COREPACK_ENABLE_DOWNLOAD_PROMPT=0
+    COREPACK_ENABLE_PROJECT_SPEC=0 (If we decide to use the global default)
+    YARN_YARN_OFFLINE_MIRROR=<absolute path to the request output directory>
+    YARN_YARN_OFFLINE_MIRROR_PRUNING=false
+    YARN_IGNORE_PATH=true (Use the corepack yarn version, not any Yarn version specified by the user)
+    YARN_IGNORE_SCRIPTS=true
+
+### Build Configuration
+The user build can be configured to use the offline mirror by configuring the following environment variables:
+
+    YARN_YARN_OFFLINE_MIRROR=<request output directory>
+    YARN_YARN_OFFLINE_MIRROR_PRUNING=false
+
+## High-Level Procedure
+ - Since we're executing Yarn commands directly, process the request in an [isolated temporary directory](https://github.com/containerbuildsystem/cachi2/blob/6953607b6ef52fd3f0bef7059d2c926767b1022b/cachi2/core/resolver.py#L41) to avoid any potential changes to the repository
+ - Determine whether to process the request with yarn v1 or yarn v2+
+ - Read the project files
+   - Read/Require package.json
+   - Read/Require yarn.lock
+     - Ensure yarn.lock is formatted for Yarn 1.x
+ - Verify the repository
+   - Ensure node_modules directory is not present
+ - Gather the SBOM Components
+   - Get the main package component data from package.json
+   - Get the workspaces from package.json and add a component for each workspace according to the workspace package.json data
+   - Parse yarn.lock with pyarn and add components for each
+     - Report missing integrity keys from yarn.lock as missing hashes
```

Comment:
Can someone more knowledgeable of Yarn v1 explain what the following is supposed to achieve and whether it's related to this problem: https://classic.yarnpkg.com/en/docs/yarnrc#toc-unsafe-disable-integrity-migration

Would it help or it's completely irrelevant?

Reply:
We report missing **_user-provided_** hashes in the SBOM with the `cachi2:missing_hash:in_file` property, so if we encounter missing hashes for any dependencies in yarn.lock we should report them as such.

I think the `unsafe-disable-integrity-migration` option was created when the integrity key was first introduced to yarn.lock several years ago. It provided a mechanism for preventing yarn from automatically updating all of the existing dependencies in a lockfile with the new integrity values (which didn't become required yarn v2 IIRC).